### ISSUE TITLE
Perf: improve UniqueCoordinateArrayFilter

### ISSUE THREAD

**Issue by jandam** [locationtech/jts#420]
UniqueCoordinateArrayFilter can be improved by using HashSet instead of TreeSet and by removing treeSet.contains
*) HashSet is faster (up to 2x) than TreeSet in most cases
*) Set::contains + Set::add can be replaced with single Set:add that return FALSE when value is already in Set.
Proposed fix:
```
public class UniqueCoordinateArrayFilter implements CoordinateFilter 
{
  /**
   * Convenience method which allows running the filter over an array of {@link Coordinate}s.
   * 
   * @param coords an array of coordinates
   * @return an array of the unique coordinates
   */
  public static Coordinate[] filterCoordinates(Coordinate[] coords)
  {
    UniqueCoordinateArrayFilter filter = new UniqueCoordinateArrayFilter();
    for (int i = 0; i < coords.length; i++) {
      filter.filter(coords[i]);
    }
    return filter.getCoordinates();
  }
  
  Set set = new HashSet();  // changed here
  List list = new ArrayList();

  public UniqueCoordinateArrayFilter() { }

  /**
   *  Returns the gathered <code>Coordinate</code>s.
   *
   *@return    the <code>Coordinate</code>s collected by this <code>CoordinateArrayFilter</code>
   */
  public Coordinate[] getCoordinates() {
    Coordinate[] coordinates = new Coordinate[list.size()];
    return (Coordinate[]) list.toArray(coordinates);
  }

  public void filter(Coordinate coord) {
    if (set.add(coord)) {  // changed here, removed exlamation
      list.add(coord);
    }
  }
}

```

**Comment by dr-jts** [comment]
Curious to know how you identified this fix.  Did you notice a performance issue with computing Convex Hulls?  Or was this flagged by an automated code style checker?

One reason for this question is that I notice a couple of potential further performance improvements:

* Use a `CoordinateSequenceFilter` instead of a `CoordinateFilter`
* Technically it is not necessary to include the vertices in polygon holes, since they should always be contained in the convex hull of the polygon shell (if it is valid).

**Comment by dr-jts** [comment]
> One reason for this question is that I notice a couple of potential further performance improvements:
> 
> * Use a `CoordinateSequenceFilter` instead of a `CoordinateFilter`
> * Technically it is not necessary to include the vertices in polygon holes, since they should always be contained in the convex hull of the polygon shell (if it is valid).

@dbaston Any thoughts on these suggestions?

**Comment by jandam** [comment]
I found this code by accident. There is no performance issue with ConvexHull. I was doing some intersection tests. I used Envelope but it covers huge area for some complex geometries. Original geometry has too many points. So I decide to use ConvexHull as original geometry approximation before final intersection test.
For example: Vltava river in Prague (~ 3000 points)
Envelope covers almost whole Prague
Convex hull covers ~10% of Prague area and has < 50 points - big speed improvement
--
Big O notation
TreeSet - O(log n) 
  - used twice (contains + add)
HashSet - O(1), O(log n) - for hash collisions after threshold
  - used once (add)
  - slightly less memory per entry

**Comment by dbaston** [comment]
@dr-jts For the convex hull application, I'd wonder about eliminating the unique filter altogether. Would the  duplicated coordinates not get tossed out in `ConvexHull::reduce` anyway? Ignoring holes makes sense and could be handled in `ConvexHull::extractCoordinates`.

**Comment by dbaston** [comment]
(`ConvexHull::reduce` method looks like another place where a `TreeSet` could be replaced, too)

**Comment by dr-jts** [comment]
> (`ConvexHull::reduce` method looks like another place where a `TreeSet` could be replaced, too)

Yes, good point.  (And there may be quite a few other places where TreeSet<Coordinate> is used and could be replaced).

**Comment by dr-jts** [comment]
> @dr-jts For the convex hull application, I'd wonder about eliminating the unique filter altogether. Would the duplicated coordinates not get tossed out in `ConvexHull::reduce` anyway? 

Yes, seems like they would.  The reduce approach was a subsequent addition to `ConvexHull`, so perhaps just didn't notice that it makes unique points redundant.

> Ignoring holes makes sense and could be handled in `ConvexHull::extractCoordinates`.

Right, that's the place to do it.

**Comment by dr-jts** [comment]
> @dr-jts For the convex hull application, I'd wonder about eliminating the unique filter altogether. Would the duplicated coordinates not get tossed out in `ConvexHull::reduce` anyway? 

It turns out that one reason for the UniqueCoordinateFilter is to handle pathological situations where there are only one or two unique points in the input.  But this could be done much more efficiently by simply scanning to check for one or two different points (and short-circuiting in the common case where there are more than two).

And it also looks like eliminating the full unique check is much faster - perhaps 6x.  So definitely worth optimizing this.

**Comment by dr-jts** [comment]
Fixed by #422
